### PR TITLE
endpoint: create separate interface for proxy

### PR DESCRIPTION
**Pull request by ianvernon**
This is waiting on https://github.com/cilium/cilium/pull/9183 to be merged.

The `Owner` interface has served as a catch-all callback into the `Daemon` which is
managing an `Endpoint`. However, the necessary functions for proxy configuration are
specific enough to warrant their own separate interface. Create a separate interface
for the functions necessary for Endpoint<->Proxy interaction, and inject this L7 proxy
into the Endpoint upon Endpoint creation / Endpoint restoration.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9157)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build got stuck: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/14894/

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Review by gandro — APPROVED** [review]

**Comment by ianvernon** [comment]
test-me-please

File: daemon/state.go
Comment (aanm) [inline daemon/state.go]: This should be here https://github.com/cilium/cilium/blob/b8aabb4de02e8f2c9e73cf633aa887e73ecbf478/daemon/state.go#L163-L168

File: pkg/endpoint/endpoint.go
Comment (aanm) [inline pkg/endpoint/endpoint.go]: Can you put this after the `ID`? I know this sounds like a weird request but sometimes it is possible to derive the endpoint ID from stack traces, works because ID is one of the first fields of this structure.

File: pkg/endpoint/policy.go
Comment (aanm) [inline pkg/endpoint/policy.go:81]: Hmm... this is not necessarily true.. check https://github.com/cilium/cilium/issues/9100 or at least comment somewhere this is being added

**Review by aanm — CHANGES_REQUESTED** [review]

File: daemon/state.go
Comment (ianvernon) [reply daemon/state.go]: @aanm the proxy is initiated after the call to `restoreOldEndpoints`:

restoreOldEndpoints is called here:

https://github.com/cilium/cilium/blob/8ff912d66aa8ec7d61cd6e7113ae29411963675e/daemon/daemon.go#L443-L446

proxy is initiated later:

https://github.com/cilium/cilium/blob/8ff912d66aa8ec7d61cd6e7113ae29411963675e/daemon/daemon.go#L517-L525

I could perhaps have a separate way to inject the proxy into the endpoints after L525?

File: pkg/endpoint/policy.go
Comment (ianvernon) [reply pkg/endpoint/policy.go:81]: I am preserving existing behavior here: the call to `e.owner.UpdateNetworkPolicy` has now been removed. The logic in the `Owner`'s `UpdateNetworkPolicy` was the following:

```
// UpdateNetworkPolicy adds or updates a network policy in the set		
 // published to L7 proxies.		
 func (d *Daemon) UpdateNetworkPolicy(e regeneration.EndpointUpdater, policy *policy.L4Policy,		
 	proxyWaitGroup *completion.WaitGroup) (error, revert.RevertFunc) {		
 	if d.l7Proxy == nil {		
 		return fmt.Errorf("can't update network policy, proxy disabled"), nil		
 	}		
 	err, revertFunc := d.l7Proxy.UpdateNetworkPolicy(e, policy, e.GetIngressPolicyEnabledLocked(),		
 		e.GetEgressPolicyEnabledLocked(), proxyWaitGroup)		
 	return err, revert.RevertFunc(revertFunc)		
 }
```

File: daemon/state.go
Comment (ianvernon) [reply daemon/state.go]: Alternatively, I can try starting the proxy before restoring endpoints. The restored endpoints aren't even exposed to the endpointmanager yet when we call `StartProxySupport`, so this should be OK.

Comment (ianvernon) [reply daemon/state.go]: Shuffled the ordering of initialization of the proxy to be before endpoints are restored.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: daemon/state.go
Comment (ianvernon) [reply daemon/state.go]: Done!

**Comment by ianvernon** [comment]
Build failed: ```22:59:41  go: creating work dir: mkdir /tmp/go-build539024766: no space left on device```

**Comment by ianvernon** [comment]
test-me-please

File: daemon/daemon_test.go
Comment (aanm) [inline daemon/daemon_test.go]: This looks like a deja vu

**Review by aanm — CHANGES_REQUESTED** [review]
All good for the last 2 commits. The first commit from this PR looks a leftover?

File: daemon/daemon_test.go
Comment (ianvernon) [reply daemon/daemon_test.go]: @aanm I updated the PR description to say that this is waiting on #9183 to be merged - the reason I did that change was for this PR to work, otherwise the proxy instantiation would fail due to the way I changed the ordering of operations for daemon bootstrap. Once #9183 is approved and merged I will drop the first commit.

**Comment by ianvernon** [comment]
test-me-please

File: daemon/daemon_test.go
Comment (ianvernon) [reply daemon/daemon_test.go]: @aanm dropped this commit and rebased against master PTAL

**Review by aanm — APPROVED** [review]

**Comment by ianvernon** [comment]
test-me-please

**Review by jrajahalme — COMMENTED** [review]
LGTM, sorry for the late review!

**Pull request merged**
